generating 2 generic closures with different generic types, instead of just 1, leads to cgen error

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.3.3 6b4fb0f.b40aa4f
OS: linux, Ubuntu 20.04.5 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz

getwd: /v/vnew
vexe: /v/vnew/v
vexe mtime: 2023-03-30 13:27:02

vroot: OK, value: /v/vnew
VMODULES: OK, value: /home/delian/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.40.0
Git vroot status: weekly.2023.13-21-g822acd78
.git/config present: true

CC version: cc (Ubuntu 10.3.0-1ubuntu1~20.04) 10.3.0
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg bug.v`
```v
fn abc[T](x T) fn(p T) T {
    return fn [x] [T](p T) T {
        return p * x
    }
}

fn main() {
    f := abc[int](12345)
    a := f(2)
    dump(a)
    //
    g := abc[u8](5)
    b := g(2)
    dump(b)
}

```

**What did you expect to see?**

a compiled program, that runs, and prints `a: 24690` and `b: 10`


**What did you see instead?**
```
==================
/tmp/v_1000/bug.18122584257121884246.tmp.c:12737: error: 'anon_fn_70a6b7e29084d475_t__T_56_T_u8' undeclared
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

I just wanted to submit a bug report about a related issue.

When using the following code everything gets correctly generated, except for the **second** closure.

```v
type ClosureHandler = fn (a string, b string)

struct ClosureStruct {
	path    string
	handler ClosureHandler
}

struct App {
	name string
}

struct Admin {
	name string
}

fn generate_closure[T](path string, global_app &T) ClosureStruct {
	return ClosureStruct{
		path: path
		handler: fn [global_app, path] [T](a string, b string) {
			println('${path}: "${a}" "${b}"')
		}
	}
}

fn main() {
	mut closures := []ClosureStruct{}

	closures << generate_closure('/app', &App{})
	closures << generate_closure('/admin', &Admin{})

	for cl in closures {
		cl.handler('a', 'b')
	}
}
```

```
/tmp/v_1000/generics.15699330998495298770.tmp.c:12703: error: 'anon_fn_3f6389a77bec19f9_string_string_331_T_main__Admin' undeclared
```

--- a/cmd/tools/vast/vast.v
+++ b/cmd/tools/vast/vast.v
@@ -581,7 +581,11 @@ fn (t Tree) anon_fn(node ast.AnonFn) &Node {
 	obj.add_terse('decl', t.fn_decl(node.decl))
 	obj.add('inherited_vars', t.array_node_arg(node.inherited_vars))
 	obj.add_terse('typ', t.type_node(node.typ))
-	obj.add('has_gen', t.bool_node(node.has_gen))
+	symbol_obj := new_object()
+	for key, val in node.has_gen {
+		symbol_obj.add_terse(key.str(), t.bool_node(val))
+	}
+	obj.add_terse('has_gen', symbol_obj)
 	return obj
 }
 

--- a/vlib/v/ast/ast.v
+++ b/vlib/v/ast/ast.v
@@ -508,7 +508,7 @@ pub mut:
 	decl           FnDecl
 	inherited_vars []Param
 	typ            Type // the type of anonymous fn. Both .typ and .decl.name are auto generated
-	has_gen        bool // has been generated
+	has_gen        map[string]bool // has been generated
 }
 
 // function or method declaration

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -204,7 +204,7 @@ fn (mut g Gen) gen_fn_decl(node &ast.FnDecl, skip bool) {
 	is_closure := node.scope.has_inherited_vars()
 	mut cur_closure_ctx := ''
 	if is_closure {
-		cur_closure_ctx = closure_ctx(node)
+		cur_closure_ctx = g.closure_ctx(node)
 		// declare the struct before its implementation
 		g.definitions.write_string(cur_closure_ctx)
 		g.definitions.writeln(';')
@@ -461,8 +461,12 @@ fn (mut g Gen) c_fn_name(node &ast.FnDecl) !string {
 
 const closure_ctx = '_V_closure_ctx'
 
-fn closure_ctx(node ast.FnDecl) string {
-	return 'struct _V_${node.name}_Ctx'
+fn (mut g Gen) closure_ctx(node ast.FnDecl) string {
+	mut fn_name := node.name
+	if node.generic_names.len > 0 {
+		fn_name = g.generic_fn_name(g.cur_concrete_types, fn_name)
+	}
+	return 'struct _V_${fn_name}_Ctx'
 }
 
 fn (mut g Gen) gen_anon_fn(mut node ast.AnonFn) {
@@ -476,7 +480,7 @@ fn (mut g Gen) gen_anon_fn(mut node ast.AnonFn) {
 		g.write(fn_name)
 		return
 	}
-	ctx_struct := closure_ctx(node.decl)
+	ctx_struct := g.closure_ctx(node.decl)
 	// it may be possible to optimize `memdup` out if the closure never leaves current scope
 	// TODO in case of an assignment, this should only call "__closure_set_data" and "__closure_set_function" (and free the former data)
 	g.write('__closure_create(${fn_name}, (${ctx_struct}*) memdup_uncollectable(&(${ctx_struct}){')
@@ -520,14 +524,18 @@ fn (mut g Gen) gen_anon_fn(mut node ast.AnonFn) {
 }
 
 fn (mut g Gen) gen_anon_fn_decl(mut node ast.AnonFn) {
-	if node.has_gen {
+	mut fn_name := node.decl.name
+	if node.decl.generic_names.len > 0 {
+		fn_name = g.generic_fn_name(g.cur_concrete_types, fn_name)
+	}
+	if node.has_gen[fn_name] {
 		return
 	}
-	node.has_gen = true
+	node.has_gen[fn_name] = true
 	mut builder := strings.new_builder(256)
 	builder.writeln('/*F*/')
 	if node.inherited_vars.len > 0 {
-		ctx_struct := closure_ctx(node.decl)
+		ctx_struct := g.closure_ctx(node.decl)
 		builder.writeln('${ctx_struct} {')
 		for var in node.inherited_vars {
 			var_sym := g.table.sym(var.typ)

--- a/vlib/v/gen/js/fn.v
+++ b/vlib/v/gen/js/fn.v
@@ -766,10 +766,14 @@ fn (mut g JsGen) fn_args(args []ast.Param, is_variadic bool) {
 }
 
 fn (mut g JsGen) gen_anon_fn(mut fun ast.AnonFn) {
-	if fun.has_gen {
+	mut fn_name := fun.decl.name
+	if fun.decl.generic_names.len > 0 {
+		fn_name = g.generic_fn_name(g.cur_concrete_types, fn_name)
+	}
+	if fun.has_gen[fn_name] {
 		return
 	}
-	fun.has_gen = true
+	fun.has_gen[fn_name] = true
 	it := fun.decl
 	cur_fn_decl := g.fn_decl
 	unsafe {

--- a/vlib/v/tests/generics_closures_with_different_generic_types_test.v
+++ b/vlib/v/tests/generics_closures_with_different_generic_types_test.v
@@ -1,0 +1,17 @@
+fn abc[T](x T) fn (p T) T {
+	return fn [x] [T](p T) T {
+		return p * x
+	}
+}
+
+fn test_generic_closures_with_different_generic_types() {
+	f := abc[int](12345)
+	a := f(2)
+	dump(a)
+	assert a == 24690
+
+	g := abc[u8](5)
+	b := g(2)
+	dump(b)
+	assert b == 10
+}

